lib/charms/alertmanager_k8s/v0/alertmanager.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+""" # AlertmanagerConsumer library
+
+This library is design to be used by a charm consuming or providing the alertmanager-k8s relation.
+"""
+
+import ops
+from ops.framework import StoredState, EventSource, EventBase
+from ops.relation import ConsumerBase, ProviderBase
+from ops.charm import CharmBase
+
+from typing import List
+import logging
+
+LIBID = "abcdef1234"  # Unique ID that refers to the library forever
+LIBAPI = 0  # Must match the major version in the import path.
+LIBPATCH = 1  # The current patch version. Must be updated when changing.
+
+logger = logging.getLogger(__name__)
+
+
+class ClusterChanged(EventBase):
+    """Event raised when an alertmanager cluster is changed.
+
+    If an alertmanager unit is added to or removed from a relation,
+    then a :class:`ClusterChanged` event is raised.
+    """
+
+    def __init__(self, handle, data=None):
+        super().__init__(handle)
+        self.data = data
+
+    def snapshot(self):
+        """Save relation data."""
+        return {"data": self.data}
+
+    def restore(self, snapshot):
+        """Restore relation data."""
+        self.data = snapshot["data"]
+
+
+class AlertmanagerConsumer(ConsumerBase):
+    """A "consumer" handler to be used by charms that relate to Alertmanager.
+
+    This consumer auto-registers relation events on behalf of the user and communicates information
+    directly via `_stored`
+    Every change in the alertmanager cluster emits a :class:`ClusterChanged` event that the
+    consumer charm can register and handle, for example:
+
+        self.framework.observe(self.alertmanager_lib.cluster_changed,
+                               self._on_alertmanager_cluster_changed)
+
+    The updated alertmanager cluster can then be obtained via the
+
+    This consumer library expect the consumer charm to register the `get_cluster_info` method.
+
+    Arguments:
+            charm (CharmBase): consumer charm
+            relation_name (str): from consumer's metadata.yaml
+            consumes (dict): provider specifications
+            multi (bool): multiple relations flag
+
+    Attributes:
+            charm (CharmBase): consumer charm
+    """
+
+    _stored: StoredState
+    cluster_changed = EventSource(ClusterChanged)
+
+    def __init__(self, charm: CharmBase, relation_name: str, consumes: dict, multi: bool = False):
+        super().__init__(charm, relation_name, consumes, multi)
+        self.charm = charm
+        self._consumer_relation_name = relation_name  # from consumer's metadata.yaml
+
+        self.framework.observe(
+            self.charm.on[self._consumer_relation_name].relation_changed, self._on_relation_changed
```

Comment:
The structure of this library is certainly improved however I am still concerned that on a relation changed event this library can emit a `ClusterChanged` event while `ConsumerBase` emits a `ProviderInvalid`. Do you see this being a possibility ? If this does happen wouldn't this result in Prometheus adding an alertmanager into its config that is and "invalid" alertmanager ?

Reply:
`ClusterChanged` event while `ProviderInvalid` is probably a possible scenario, but I think there is no wrong with it: 
1. alertmanager having an IP address is independent of the provider being `ready` or `unready`. Moreover, taking into account `ConsumerEvents` would not be easy, and we also do not have a pattern for it yet, afaik.
2. if an alertmanager unit joins but the provider did not set `ready` yet, it's ok for a consumer charm to have the ip address.
3. if an alertmanager unit leaves and the provider is unready, it makes sense for the consumer to remove it.

Reply:
In invalid provider has to do with a version or type incompatibility with the consumer, in dependent of provider is ready or not. If you agree `ClusterChanged` can be raised when provider is invalid this means that Prometheus config will have alertmanager that it is version incompatible with. Is this a bug ?

Reply:
oh ok I see, so `ProviderInvalid` would be emitted once and that's it right?
In that case:
1. "version mismatch", as far as alertmanager and say prometheus are concerned, is probably an "unsupported api" error, e.g. prometheus tries to access `/api/v2`, which doesn't exist for the version of alertmanager in use. But this is hypothetical and probably the responsibility of prometheus: prometheus should query alertmanager's `/api/status/version` and use the correct alertmanager api endpoints.
2. Taking into account ProviderInvalid would not be easy and we do not have a pattern for it yet, afaik.

Reply:
You are saying that the Prometheus charm should be responsible for ensuring it can deal with an incompatible version of an Alertmanager charm, despite the fact that

1. The Prometheus charm (the consumer) specifies the minimum Alertmanager version using a consumers clause.
2. The Provider Consumer library raises a ProviderInvalid event.

I do not see what then is the point of the Provider Consumer library. Do you ? 

With regard to ready and unready. Can't it happen that the Provider (Alertmanager) is ready but still incompatible with Prometheus ?

Reply:
Captured in #9, will address in next PR.